indexing-service/src/main/java/io/druid/indexing/overlord/TaskLockbox.java

```diff
@@ -310,6 +351,146 @@ public TaskLock lock(final Task task, final Interval interval) throws Interrupte
+    boolean canAcquire = true;
+    final Map<String, TaskLock> tasksToPreempt = new HashMap<>();
+    for (TaskLockPosse taskLockPosse : foundPosses) {
+      if(lockPriority <= taskLockPosse.getTaskLock().getPriority() || taskLockPosse.getTaskLock().isUpgraded()) {
+        canAcquire = false;
+        tasksToPreempt.clear();
+        break;
+      } else {
+        // input lockPriority is greater than lockPriority of this TaskLockPosse
+        // in case we acquire a lock revoke the taskLocks for the tasks in this TaskLockPosse
+        for (String taskId : taskLockPosse.getTaskIds()) {
+          tasksToPreempt.put(taskId, taskLockPosse.getTaskLock());
+        }
+      }
+    }
+    return new Pair<>(canAcquire, tasksToPreempt);
+  }
+
+  /**
+   * Unlock TaskLocks
+   * */
+  private void revokeTaskLocks(Map<String, TaskLock> tasksToPreempt) {
+    // Revoke lock of tasks which has lower priority and conflicts with the lock of new task
+    for (Map.Entry<String, TaskLock> entry : tasksToPreempt.entrySet()) {
+      Task task = taskStorage.getTask(entry.getKey()).get();
+      log.info(
+          "Revoking task lock with datasource=[%s] and interval=[%s] for task [%s] with priority=[%s] groupId=[%s]",
+          entry.getValue().getDataSource(),
+          entry.getValue().getInterval(),
+          task.getId(),
+          task.getLockPriority(),
+          task.getGroupId()
+      );
+      unlock(task, entry.getValue().getInterval());
+    }
+  }
+
+  /**
+   * Given GroupId, DataSource, Interval, Version and Priority create a new TaskLockPosse
+   * Add the TaskLockPosse to the in-memory data structure keeping track of all the TaskLockPosse sorted by interval for a datasource
+   * */
+  private TaskLockPosse createNewTaskLockPosse(String groupId, String dataSource, Interval interval, String version, Integer priority) {
+    giant.lock();
+    try {
+      if (!running.containsKey(dataSource)) {
+        running.put(dataSource, new TreeMap<Interval, TaskLockPosse>(Comparators.intervalsByStartThenEnd()));
+      }
+      TaskLockPosse taskLockPosse = new TaskLockPosse(
+          new TaskLock(
+              groupId,
+              dataSource,
+              interval,
+              version,
+              priority,
+              null
+          )
+      );
+      running.get(dataSource).put(interval, taskLockPosse);
+      log.info("Created new TaskLockPosse: %s", taskLockPosse);
+      return taskLockPosse;
+    } finally {
+      giant.unlock();
+    }
+  }
+
+  /**
+   * Toggle the taskLock for this task
+   *
+   * @param task             task that has this lock
+   * @param interval         interval for the lock
+   * @param taskLockToggle   upgrade or downgrade the lock depending on this parameter
+   *
+   * @return true if the taskLock was toggled, false otherwise
+   */
+  public boolean toggleLock(Task task, Interval interval, TaskLockToggle taskLockToggle) {
```

Comment:
yes, here it is trying to set a state for the lock depending on TaskLockToggle and it may not actually toggle the lock if the lock is already in desired state. Do you want me to change the name of method ?